add-privileged-users-in-room

GoodGuyMarco opened the pull request:
This PR is required by https://github.com/matrix-org/matrix-react-sdk/pull/9596.

Notes: Enable multiple users' power levels to be set at once

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [ ] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

Signed-off-by: Marco Bartelt marco.bartelt.ext@bwi.de


<!-- CHANGELOG_PREVIEW_START -->
---
Here's what your changelog entry will look like:

## ✨ Features
 * Enable multiple users' power levels to be set at once ([\#2892](https://github.com/matrix-org/matrix-js-sdk/pull/2892)). Contributed by @GoodGuyMarco.<!-- CHANGELOG_PREVIEW_END -->

robintown reviewed (approved):
Looks fine to me. Sorry this took a while to be reviewed!

robintown commented:
@GoodGuyMarco Our CI seems to be stuck. Could you please update your branch to restart it?

GoodGuyMarco commented:
@robintown @t3chguy Thank you for your review! I've just rebased this PR against `develop`. I also added a unit test covering the new changes. Please look into that aswell 🕵️

t3chguy commented on src/client.ts:
```diff
@@ -3795,29 +3795,38 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
         event: MatrixEvent,
     ): Promise<ISendEventResponse> {
-        let content = {
+        type EventContent = {
```
the utils.deepCopy later on breaks this type, given it'll include other fields like `events`, `state_default`, etc. Once the typing for `deepCopy` is fixed this would no longer comply.

GoodGuyMarco replied:
I see, any idea how to satisfy the strict type checker? I don't get why it fails. Locally it does not have any TypeScript issues in strict mode.

t3chguy replied:
Make sure to also run `--noImplicitAny` given we (wrongly) specify that as `false` in tsconfig

GoodGuyMarco replied:
Ah, thanks for the hint 💡  Is this a better solution: https://github.com/matrix-org/matrix-js-sdk/pull/2892/commits/763532f84a5774b90ce62208e8908545ec16ed7c?

t3chguy commented on src/client.ts:
```diff
@@ -3795,29 +3795,36 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
     ): Promise<ISendEventResponse> {
+        type UsersWithPowerLevel = Record<string, number>;
         let content = {
             users: {},
```
Wonder if you could do `users: {} as Record<string, number>,` and then not need the type assertions later

GoodGuyMarco replied:
Yes, that seems to work aswell 👍

t3chguy commented on spec/integ/matrix-client-methods.spec.ts:
```diff
@@ -1337,27 +1337,38 @@ describe("MatrixClient", function() {
-                matrix_server_name: "server_name",
-                expires_in: 12345,
-            };
+    describe("setPowerLevel", () => {
```
It looks like you replaced an existing unrelated test?

GoodGuyMarco replied:
Yes, there were two completely identical tests, see: https://github.com/matrix-org/matrix-js-sdk/blob/develop/spec/integ/matrix-client-methods.spec.ts#L1316

t3chguy reviewed (approved):
Looks sane to me, thanks for iterating!!
